### DISCUSSION TITLE
No CSS being loaded and base href on http (self hosted)

### DISCUSSION THREAD

**Discussion by cyavu** [firefly-iii/firefly-iii#9569] [Firefly III] [answered]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [x] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/explanation/more-information/what-its-not/)
- [x] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been requested before.

### Description

So there have been multiple issues over the past years of users reporting that no CSS is being loaded or the base href being on 'http' instead of 'https'. The help page also suggests ([here](https://docs.firefly-iii.org/references/faq/docker/#all-links-refer-to-http-so-forms-and-charts-dont-work)) that setting TRUSTED_PROXIES to * should fix this, but this doesn't apply to self hosted instances where that parameter does nothing.

Not having CSS loaded and the base href being on http will prevent your browser you from logging in, flagging it as insecure.

### Solution

The solution is to add

```
if($this->app->environment('production')) {
    \URL::forceScheme('https');
}
```

to _app/Providers/AppServiceProvider.php_ so the code looks like the following:

```
class AppServiceProvider extends ServiceProvider
{
    /**
     * Bootstrap any application services.
     */
    public function boot(): void
    {
if($this->app->environment('production')) {
    \URL::forceScheme('https');
}
```

Remove the **if statement** if your env is set to 'testing' or you want to load it unconditionally. Otherwise, change your env to production.

You could, optionally, add a ASSET_URL at the top of your .env and that alone will fix the CSS issues, not the base href being loaded on http.

Perhaps the devs could include this, since you'll have to repeat these steps after every update. **The same applies to the data-importer**.

**Comment by JC5** [comment]
Hi, thanks for opening an issue.

This will not happen, thanks for the suggestion though. A correctly configured web server will have no problem serving https urls.

**Comment by cyavu** [comment]
> Hi, thanks for opening an issue.
> 
> This will not happen, thanks for the suggestion though. A correctly configured web server will have no problem serving https urls.

Hi JC5,

Thank you for your reply.

It did with me, though. I've been using Firefly for a year now and this started with some versions earlier, can't really tell when exactly. I assume most people just run Firefly within Docker and never had to deal with this.

**Comment by JC5** [comment]
Forcing the protocol is a bit of a hack and it's an opening for more problems down the line. What's your setup like?

**Reply by cyavu** [reply]
> Forcing the protocol is a bit of a hack and it's an opening for more problems down the line. What's your setup like?

True, hence I am open to alternatives as this is not the nicest way of doing things. I'm running plain Apache + MySQL with no nginx.

**Comment by atb00ker** [comment]
I just ran into the same error here is some details:
Happens on `APP_ENV=production`

1. If I set `TRUSTED_PROXIES=**`, it works.
2. If I set `TRUSTED_PROXIES` to `bank.example.com` or `https://bank.example.com` , it sets the baseurl in html body to `http://...` causing the issue.

Here is my nginx block:

```
    ...
    http2           on;
    listen 443      ssl;
    listen [::]:443 ssl;
    server_name     bank.example.com;

    # upstream
    set $upstream_firefly firefly:8080;

   ...
    location / {
        proxy_pass http://$upstream_firefly;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;
    }
```


Am I doing something stoopid or is it an error? 🤔

**Reply by JC5** [reply]
Yeah don't do that. The `TRUSTED_PROXIES` environment variable tells the Docker container which reverse proxy to trust. It only accepts `*` (or `**`) or an IP address.

**Reply by atb00ker** [reply]
Super. Worked, I'll go learn to read instructions properly now. 😂

**Comment by cyavu** [comment]
On the latest release (6.2.4), the issue still persists. Now you only need to add the following:
```\URL::forceScheme('https');```

![afbeelding](https://github.com/user-attachments/assets/e37d6b0f-322c-478a-8b5d-cecb26a20dca)

**Reply by JC5** [reply]
Hey, thanks for your suggestion. Please don't do this. It's a cheap hack that won't solve the problem.

**Reply by cyavu** [reply]
I don't understand. My installation simply does not work without enforcing this setting. Mind you that this only started happening a few releases ago and that I've been running the self-hosted package for over a year.

Yes, it is a cheap hack, but also yes it solves my problems and I have yet to find a permanent solution.

**Reply by cyavu** [reply]
By the way, the same is occuring on the data-importer.

**Reply by JC5** [reply]
Sure, but that means that your reverse proxy is not configured correctly. There are several solutions presented in this thread.

**Reply by cyavu** [reply]
> S

There is no reverse proxy. It's plain Apache + MySQL. I'm sure it all works fine in the Docker images but not the self-hosted variant (see title and earlier comments made by me).

**Reply by JC5** [reply]
Mmmm. Still though. My own development setup is also self hosted and works over https. Would it work if I shared my configuration later?

**Reply by cyavu** [reply]
Yes, please!

**Reply by JC5** [reply]
Alright here it is. This includes both http and https (and does NOT forward to https) because I had issues with self-signed certificates in the past.

```
server {
        listen 80;
        location ~ \.php$ {
                try_files $uri =404;
                fastcgi_pass unix:/var/run/php/php8.4-fpm.sock;
                fastcgi_index index.php;
                fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
                include fastcgi_params;
                fastcgi_read_timeout 300;
        }

        large_client_header_buffers 8 32k;

        root /ff3/release/firefly-iii/public;
        client_max_body_size 300M;
        index index.php index.html index.htm;
        server_name firefly.sd.internal;

        location / {
                try_files $uri $uri/ /index.php?$query_string;
                autoindex on;
                sendfile off;
        }
}

server {
        listen 443 ssl;
        listen [::]:443 ssl;

        large_client_header_buffers 8 32k;

        root /ff3/release/firefly-iii/public;
        client_max_body_size 300M;
        index index.php index.html index.htm;
        server_name firefly.sd.internal;

        location / {
                try_files $uri $uri/ /index.php?$query_string;
                autoindex on;
                sendfile off;
                fastcgi_hide_header X-Powered-By;
        }



        location ~ \.php$ {
                try_files $uri =404;
                fastcgi_pass unix:/var/run/php/php8.4-fpm.sock;
                fastcgi_index index.php;
                fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
                include fastcgi_params;
                fastcgi_read_timeout 300;
                include /etc/nginx/001-firefly-release-cors.conf;
        }


        ssl_certificate /etc/nginx/tls/wildcard.sd.internal.crt;
        ssl_certificate_key /etc/nginx/tls/wildcard.sd.internal.key;
        include /etc/nginx/tls/tls.conf;
}
```

You see references to two configuration files, so here they are.

**001-firefly-release-cors.conf**. I believe this file is NOT necessary to make it work, but here it is regardless.

```
set $cors '';
set $cors_allowed_methods 'OPTIONS, HEAD, GET, PUT, DELETE';

if ($http_origin ~ '^https://firefly-dev.sd.internal:8080$') {
        set $cors 'origin_matched';
}

if ($http_origin ~ '^https://localhost:8080$') {
        set $cors 'origin_matched';
}

add_header X-Origin $http_origin always;

# Preflight requests
if ($request_method = OPTIONS) {
    set $cors '${cors} & preflight';
}

if ($cors = 'origin_matched') {
    add_header Access-Control-Allow-Origin $http_origin always;
    add_header Access-Control-Allow-Methods $cors_allowed_methods always;
    add_header Access-Control-Allow-Credentials 'true' always;
    add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,x-xsrf-token,x-csrf-token' always;
}

if ($cors = 'origin_matched & preflight') {
    add_header Access-Control-Allow-Origin $http_origin always;
    add_header Access-Control-Allow-Methods $cors_allowed_methods always;
    add_header 'Access-Control-Allow-Headers' 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,x-xsrf-token,x-csrf-token';
    add_header Access-Control-Allow-Credentials 'true' always;
    add_header Content-Type text/plain always;
    add_header Content-Length 0 always;
    return 204;
}
```

tls.conf. It should NOT be necessary to set these parameters, but you never know. Keep in mind that `ssl_dhparam` points to a custom file (the command to generate it is right there). 

```
# This file contains important security parameters. If you modify this file
# manually, Certbot will be unable to automatically provide future security
# updates. Instead, Certbot will print and log an error message with a path to
# the up-to-date file that you will need to refer to when manually updating
# this file.

ssl_protocols TLSv1.2 TLSv1.3;# Requires nginx >= 1.13.0 else use TLSv1.2
ssl_prefer_server_ciphers on;
ssl_dhparam /etc/nginx/tls/ssl-dhparams.pem; # openssl dhparam -out /etc/nginx/tls/ssl-dhparams.pem 4096
ssl_ciphers EECDH+AESGCM:EDH+AESGCM;

ssl_ecdh_curve secp384r1; # Requires nginx >= 1.1.0
ssl_session_timeout  10m;
ssl_session_cache shared:SSL:10m;
ssl_session_tickets off; # Requires nginx >= 1.5.9
ssl_stapling on; # Requires nginx >= 1.3.7
ssl_stapling_verify on; # Requires nginx => 1.3.7
resolver 1.1.1.1 8.8.8.8 valid=300s;
resolver_timeout 5s;

add_header X-Frame-Options DENY;
add_header X-Content-Type-Options nosniff;
add_header X-XSS-Protection "1; mode=block";
add_header X-Robots-Tag none;
add_header X-Download-Options noopen;
add_header X-Permitted-Cross-Domain-Policies none;
add_header Referrer-Policy no-referrer;
fastcgi_hide_header X-Powered-By;
```

Let me know if this helps!

**Reply by cyavu** [reply, marked as answer]
Hi JCS,
I appreciate the time and effort you're putting into this, thank you! However, the config you've sent is nginx config and I'm not using nginx. You did convince me this is related to my config, so I've done some digging.

So instead of only forcing every http request to https within the <VirtualHost *:80> block, I now also enforce https to all incoming https requests as well within the <VirtualHost *:443> config, which fixes the issue:

```
RewriteEngine On 
RewriteCond %{HTTPS} !=on 
RewriteRule ^/?(.*) https://%{SERVER_NAME}/$1 [R,L]
```

This is not a Firefly issue, but a config issue on my end. However, normally you would not need this route since Laravel is still using http under water. Hopefully someone will find this useful in the future.
Thanks again JCS.
